hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/s2sjoin/SqlSTSJoinFaultToleranceStressTest.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2021 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.s2sjoin;
+
+import com.hazelcast.jet.JetService;
+import com.hazelcast.jet.Job;
+import com.hazelcast.jet.kafka.impl.KafkaTestSupport;
+import com.hazelcast.jet.sql.SqlTestSupport;
+import com.hazelcast.jet.sql.impl.connector.kafka.KafkaSqlConnector;
+import com.hazelcast.map.IMap;
+import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlService;
+import com.hazelcast.test.HazelcastParametrizedRunner;
+import com.hazelcast.test.HazelcastSerialParametersRunnerFactory;
+import com.hazelcast.test.annotation.NightlyTest;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+import org.junit.runners.Parameterized.UseParametersRunnerFactory;
+
+import java.io.IOException;
+
+import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_KEY_FORMAT;
+import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_FORMAT;
+import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParametrizedRunner.class)
+@UseParametersRunnerFactory(HazelcastSerialParametersRunnerFactory.class)
+@Category({NightlyTest.class, ParallelJVMTest.class})
+public class SqlSTSJoinFaultToleranceStressTest extends SqlTestSupport {
```

Comment:
The test is not sufficient. We must assert exact contents of the result, not just the count. Also, we shouldn't write to IMap where keys get overwritten and we can't assert that there were no duplicates. We should write to IList.

Test variants should be `restart_graceful` and `restart_forceful`, perhaps in combination with at-least-once and ex-once. With restart_graceful+exactly_once, there must be no duplicates in the output. In other test cases duplicates are possible.

Also, we must not insert the source data at once, but insert them continuously over longer time period, so that we have a chance to actually restart a couple of times during that time and stress the processor.

We should also use the outer join, since this is the trickier case. We don't cover FT with outer joins in the unit test, because the results are not exactly the same after a restart, but I _hope_ they are equivalent, the stress test will check that.